File: v1/kotlin/binary/src/main/kotlin/Binary.kt
object Binary {
    fun isValidChar(c: Char): Boolean = c in listOf('0', '1')

    fun charToDecimal(c: Char): Int = c - '0'

    fun toDecimal(input: String): Int =
        if (input.all { isValidChar(it) })
            input.fold(0) { acc, c -> acc * 2 + charToDecimal(c) }
        else
            0
}

File: v1/csharp/binary/Binary.cs
using System.Linq;

public class Binary
{
    private const int InvalidBinaryDecimalValue = 0;

    public static int ToDecimal(string binaryString)
    {
        return CalculateDecimalValue(binaryString);
    }

    private static int CalculateDecimalValue(string binaryString)
    {
        if (IsValidBinaryString(binaryString))
        {
            return binaryString.Aggregate(0, (acc, digit) => acc * 2 + digit - '0');
        }

        return InvalidBinaryDecimalValue;
    }

    private static bool IsValidBinaryString(string binary)
    {
        return binary.All(c => c == '0' || c == '1');
    }
}